### PR TITLE
Automatically pick up changes to data models definitions

### PR DESCRIPTION
**Pull request by murphyke**
This is an implementation of option 3 in issue #51 (webhook from the data-models repo), with the addition of a periodic refresh.

The "Caching data from the data models service" section in `README.md` discusses shortcomings with this approach.

It's an improvement over the status quo, but I'm not really happy with it because of the extra complexity of messing with the threading-based timer and the fact that prompt termination of threads during shutdown is not addressed.  I tested code to call `cancel()` on the timers, but that doesn't help; greater thread foo would be required. I just hate to go down that rabbit-hole when the option of not caching anything at all would be simple and adequate for any imaginable usage load.

Another thing that occurred to me is that the whole reason for the partial caching in the first place is that the data-models-service `/models?format=json` endpoint returns way more data than is returned by `/models?format=html`. I guess that's water under the bridge now, but it does seem that there should be a json endpoint that returns what `/models?format=html` does, i.e. just a list of models and versions.

### CODE REVIEW

**Comment by murphyke** [comment]
@bruth @aaron0browne P.S. I haven't fixed the tests yet.

**Comment by murphyke** [comment]
The tests are no longer broken, although there are no tests for the new features.

File: dmsa/models.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    try:
+        with open(MODELS_FILE, mode='w') as f:
+            try:
+                pickle.dump(models, f)
+            except pickle.PicklingError as e:
+                logging.error('pickling data models: {}'.format(e))
+                raise
+    except IOError as e:
+        logging.error('opening {} for writing: {}'.format(MODELS_FILE, e))
+        raise
```
Comment (bruth) [inline dmsa/models.py]: Exceptions can be stacked. Most context managers (for `with`) don't swallow exceptions.

```python
try:
    ...
except pickle.PicklingError as e:
    ...
except IOError as e:
    ...
```

But a more general question, what is the value of logging these errors? Where do these log statements go, can they be viewed and/or alerted, and are they actionable?

File: dmsa/service.py [hunk 2]
```diff
@@ -1,13 +1,57 @@
...
+    t = threading.Timer(delay,
+                        refresh_data_models_template_and_reschedule,
+                        args=[delay if reschedule else None]
+                        )
+    t.start()
+
+
+@webhook.hook(event_type='issues')
+def webhook_route(data):
+    print 'webhook'
```
Comment (bruth) [inline dmsa/service.py]: Remove `print` statement.

**Review by bruth — CHANGES_REQUESTED** [review]
I have a slightly _radical_ idea that is just thinking more long term if these services are going to continue to evolve and be used.. why not just have the data models service subsume the endpoints and have it deal with calling out to a local Python service. That way it can POST the new data which would update the cache, the Python process should be single process so no locking needs to happen, etc. There may be other factors I am not considering @aaron0browne? A migration path would be to setup redirects from the DMSA endpoints to the corresponding DMS endpoint that is backed by the new integration.

File: dmsa/models.py (hunk 1 above)
Comment (murphyke) [reply dmsa/models.py]: We should probably adopt the convention that all of our exceptions and logging should go to "the" dbhi sentry server unless there is something more specific arranged. Is there a communal Sentry server?  I was thinking we could add Sentry support in a subsequent PR.

Comment (murphyke) [reply dmsa/models.py]: And thanks for the reminder about batching the exception handling.

File: dmsa/service.py [hunk 3]
```diff
@@ -111,7 +161,8 @@ def create_erd_route(model, version):
     ext = request.args.get('format') or 'png'
 
     filename = '{0}_{1}_dms_{2}_dmsa_{3}.{4}'.format(
-        model, version, app.config['service_version'], __version__, ext)
+        model, version, get_service_version(app.config['service']), __version__,
```
Comment (gracecodes144) [inline dmsa/service.py]: Can the `get_service_version` call be moved into `refresh_data_models_template`, instead?

**Review by gracecodes144 — APPROVED** [review]
I only have the one code comment. I hear your unhappiness with this fix @murphyke and also your desire for a more radical change @bruth and I think you're both right. So, what if we go ahead with this because it works and fixes the immediate problem and also schedule a more complete review of the data models ecosystem soon, wherein we consider more radical changes?

**Comment by bruth** [comment]
@aaron0browne 👍 I agree with that.

**Review by bruth — APPROVED** [review]

File: dmsa/service.py (hunk 3 above)
Comment (murphyke) [reply dmsa/service.py]: I'll check it out. Another thing about service version is that it is sent with every response, so strictly speaking it's not needed to make a separate request for that.

Comment (murphyke) [reply dmsa/service.py]: But I don't really feel like doing the refactorings in utility.py and changing the unit tests, so I'll just do as you suggest.

**Comment by murphyke** [comment]
I noticed that the original code was writing temp files to flask's `instance_path` directory, so I changed the lock/cache code to use that also.  The CI test had been failing when trying to write the lock file, so maybe this will fix that problem.

I also added a forgotten dependency.

**Comment by murphyke** [comment]
@aaron0browne The CircleCI build is failing with `No such file or directory: 'requirements.txt'`. Not sure what's going on.

```
Failing command: tox 
Exit code: 1 
Output:
GLOB sdist-make: /home/ubuntu/data-models-sqlalchemy/setup.py
python create: /home/ubuntu/data-models-sqlalchemy/.tox/python
python inst: /home/ubuntu/data-models-sqlalchemy/.tox/dist/dmsa-0.5.10.zip
ERROR: invocation failed (exit code 1), logfile: /home/ubuntu/data-models-sqlalchemy/.tox/python/log/python-1.log
ERROR: actionid: python
msg: installpkg
cmdargs: ['/home/ubuntu/data-models-sqlalchemy/.tox/python/bin/pip', 'install', '/home/ubuntu/data-models-sqlalchemy/.tox/dist/dmsa-0.5.10.zip']

Processing ./.tox/dist/dmsa-0.5.10.zip
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-gIFaHn-build/setup.py", line 11, in <module>
        with open('requirements.txt') as f:
    IOError: [Errno 2] No such file or directory: 'requirements.txt'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-gIFaHn-build/

python installed: 
___________________________________ summary ____________________________________
ERROR:   python: InvocationError: /home/ubuntu/data-models-sqlalchemy/.tox/python/bin/pip install /home/ubuntu/data-models-sqlalchemy/.tox/dist/dmsa-0.5.10.zip (see /home/ubuntu/data-models-sqlalchemy/.tox/python/log/python-1.log)

tox returned exit code 1
```

**Comment by murphyke** [comment]
OK, I see now that has nothing to do with CircleCI, per se. `tox` at the command line behaves the same way.

**Comment by murphyke** [comment]
@aaron0browne Adding `requirements.txt` to MANIFEST.in fixes the tox problem. Man, I totally do not understand Python packaging. I'd better read up a bit.

**Comment by murphyke** [comment]
CircleCI build works now, FWIW.

**Comment by gracecodes144** [comment]
FWIW, I think Python packaging is a mess and I don't understand it either... Are you still working on more changes or is this ready to go?

**Comment by murphyke** [comment]
@aaron0browne This is ready to go.

**Comment by gracecodes144** [comment]
OK! Version update? I believe it's in `dmsa/__init__.py`. Maybe increase the minor version and set the release level to beta. Then after we deploy and make sure its working we can increase to final directly on the master branch. Sound good?

**Comment by murphyke** [comment]
Right, OK

File: README.md [hunk 4]
```diff
@@ -229,18 +229,4 @@ The caching is also not very intelligent in that there is no coordination betwee
...
-```
-
-Deploy the service to AWS. You will be prompted for the docker image tag you want to deploy.
-
-```
-cd ci
-terraform plan
-terraform apply
-```
+Deployment of the DBHi `data-models-sqlalchemy.research.chop.edu` service is managed via the internal devops/data-models-sqlalchemy repo.
```
Comment (gracecodes144) [inline README.md]: Thanks for catching this!

**Comment by murphyke** [comment]
@aaron0browne So how do you want to proceed with this? Should I merge this, and you will do a test deployment?

**Comment by gracecodes144** [comment]
@burrowse What stage of the PEDSnet data cycle are you in right now? Deploying and testing would be easiest if we could cause a slight outage in the service without bothering anyone. Is that the case right now?

**Comment by burrowse** [comment]
@aaron0browne This week is an active ETL week for sites as the submission is due next Monday. I would imagine if there was a way for the testing to be done after hours it would be okay or we can make an announcement for the service to be unavailable during a certain time this week.

**Comment by gracecodes144** [comment]
Given that, maybe its worth the extra effort to test on a separate instance of the service. @murphyke can you think of anything that would stop us from running an integration test on a dummy data models repo using test instances of the data models service and data models sqlalchemy? I believe all the relevant configuration is easily modified. The test instances won't have SSL, but I don't think that will be a problem, do you?

**Comment by murphyke** [comment]
No, I don't think that would be a problem.

**Comment by murphyke** [comment]
@aaron0browne I think you suggested merging this PR and then building and pushing the Docker image based on that for integration testing. I will do that this morning, per your devops cookbook docs for [DMS](https://github.research.chop.edu/devops/data-models-service#cookbook) and [DMSA](https://github.research.chop.edu/devops/data-models-sqlalchemy#cookbook). Then, assuming testing works, I'll change the DMSA version from beta to final on master, build and push the final image, and deploy into production ....

**Pull request merged**